rpc/send.go

```diff
@@ -154,121 +143,92 @@ func Send(opts Options, method string, addrs []net.Addr, getArgs func(addr net.A
-			if args == nil {
-				helperChan <- util.Errorf("nil arguments returned for client %s", clients[index].Addr())
-				continue
-			}
-			reply := getReply()
+	sendFn := func(addr net.Addr) {
+		// NewClient might block while connecting, but connecting in parallel
+		// blows up the tests.
+		client, err := NewClient(addr, context)
+		if err != nil {
+			done <- &rpc.Call{Error: newRPCError(util.Errorf("connection to %s failed: %s", addr, err))}
+			return
+		}
+
+		if args := getArgs(addr); args != nil {
 			if log.V(2) {
-				log.Infof("%s: sending request to %s: %+v", method, clients[index].Addr(), args)
+				log.Infof("%s: sending request to %s: %+v", method, addr, args)
 			}
-
-			trace.Event(fmt.Sprintf("sending to %s", clients[index].Addr()))
-			go sendOneFn(clients[index], opts.Timeout, method, args, reply, helperChan)
+			trace.Event(fmt.Sprintf("sending to %s", addr))
+			go sendOneFn(client, opts.Timeout, method, args, getReply(), done)
+		} else {
+			done <- &rpc.Call{Error: newRPCError(util.Errorf("nil arguments returned for client %s", addr))}
 		}
-		// Wait for completions.
+	}
+
+	// Start clients up to opts.N.
+	head, tail := orderedAddrs[:opts.N], orderedAddrs[opts.N:]
+	for _, addr := range head {
+		sendFn(addr)
+	}
+
+	var replies []interface{}
+	var errors, retryableErrors int
+
+	// Wait for completions.
+	for len(replies) < opts.N {
 		select {
-		case r := <-helperChan:
-			switch t := r.(type) {
-			case error:
+		case call := <-done:
+			if err := call.Error; err != nil {
+				if log.V(1) {
+					log.Warningf("%s: error reply: %s", method, err)
+				}
+
 				errors++
-				if retryErr, ok := t.(retry.Retryable); ok && retryErr.CanRetry() {
+				if retryErr, ok := err.(retry.Retryable); ok && retryErr.CanRetry() {
 					retryableErrors++
 				}
-				if log.V(1) {
-					log.Warningf("%s: error reply: %+v", method, t)
-				}
-				remainingNonErrorRPCs := len(clients) - errors
-				if remainingNonErrorRPCs < opts.N {
+
+				if remainingNonErrorRPCs := len(addrs) - errors; remainingNonErrorRPCs < opts.N {
 					return nil, SendError{
-						errMsg:   fmt.Sprintf("too many errors encountered (%d of %d total): %v", errors, len(clients), t),
+						errMsg:   fmt.Sprintf("too many errors encountered (%d of %d total): %v", errors, len(clients), err),
 						canRetry: remainingNonErrorRPCs+retryableErrors >= opts.N,
 					}
 				}
 				// Send to additional replicas if available.
-				if N < len(clients) {
+				if len(tail) > 0 {
 					trace.Event("error, trying next peer")
-					N++
+					sendFn(tail[0])
+					tail = tail[1:]
 				}
-			default:
-				successes++
+			} else {
 				if log.V(2) {
-					log.Infof("%s: successful reply: %+v", method, t)
-				}
-				replies = append(replies, t)
-				if successes == opts.N {
-					return replies, nil
+					log.Infof("%s: successful reply: %+v", method, call.Reply)
 				}
+
+				replies = append(replies, call.Reply)
 			}
 		case <-time.After(opts.SendNextTimeout):
```

Comment:
Preexisting issue: the time.After should be a variable that lives outside the loop so we don't reset the clock each time a response comes in.